### ISSUE TITLE
install_requires is ignored

### ISSUE THREAD

**Issue by bsipocz** [astropy/ci-helpers#337]
`install_requires` seems to be ignored during travis testing, we need to figure out what can be done in `ci-helpers`

e.g. a build for astroML with python3.4 ignores that scikit-learn 0.18 is required in `setup.py` and happily installes a 0.16 version:
https://travis-ci.org/astroML/astroML/jobs/450658367

**Comment by bsipocz** [comment]
@drdavella or @astrofrog  - do you have any good ideas how this can be fixed? I suspect this can be critial for other affiliates, too.

**Comment by drdavella** [comment]
This line [here](https://travis-ci.org/astroML/astroML/jobs/450658367#L1154) makes me think that either Py34 is too old to support this option, or the default version of `setuptools` that comes with Py34 is too old, so you need to upgrade explicitly.

**Comment by bsipocz** [comment]
Oh, that's a very likely explanation, thanks. The pip version was also ancient.

**Comment by bsipocz** [comment]
This is also an issue on LTS astropy testing: https://github.com/astropy/astropy/pull/8173

**Comment by bsipocz** [comment]
Weird. ci-helpers ignores the `install-requires` while installing the dependencies from the lists. But when the package is installed setuptools realizes this and downgrades to <4.0.0: https://travis-ci.org/astropy/astropy/jobs/458990301#L927

yet when running the tests, the conda installed version is picked up not the one from the eggs.


(as a workaround I'll

**Comment by drdavella** [comment]
This is mostly anecdotal, but for some reason I started to wonder whether `install_requires` is even supported by `pip` at all anymore. Maybe this is because `pyproject.toml` is the recommended solution going forward? Maybe @Cadair knows? I sure wish this stuff were better documented.

**Comment by Cadair** [comment]
Sorry, I have no idea :grin:

**Comment by astrofrog** [comment]
@drdavella - ``install_requires`` is definitely still supported. The ``pyproject.toml`` is just for **build**-time-dependencies.

**Comment by bsipocz** [comment]
it doesn't interact, the point was that it would be nice if it would take into account e.g. pytest version limitation or other stuff packages define in there rather than also setting the same thing in the travis matrix manually.

**Comment by bsipocz** [comment]
(and indeed the install part takes those limitations into account, but by that time the "new" old versions will be only installed as egg, and then they are not actually used during the testing)

**Comment by astrofrog** [comment]
So one approach I've started taking for some packages is to define the bare dependencies with ``install_requires``, and then also list the full dependencies with an ``'all'`` entry in ``extras_requires``, then to use ci-helpers to set up the basic Python version but then to just do ``pip install -e .`` or ``pip install -e .[all]``or ``pip install -e [test,all]`` before running the tests to install the relevant dependencies.

I don't think we should hack ``ci-helpers`` to try and read in ``install_requires`` necessarily. Instead, it might be worth thinking about whether the workflow I just described is something that would actually work better in future now that conda is so slow and complex. This also has the benefit of making sure dependencies are properly specified in ``setup.py`` and avoids duplication in the CI config files.
